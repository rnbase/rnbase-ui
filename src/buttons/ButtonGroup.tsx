import React, { useEffect, useState } from 'react';
import {
  Animated,
  Easing,
  Image,
  ImageSourcePropType,
  ImageStyle,
  LayoutChangeEvent,
  StyleProp,
  StyleSheet,
  Text,
  TextStyle,
  TouchableOpacity,
  TouchableOpacityProps,
  View,
  ViewStyle,
} from 'react-native';

import { Themed, Theme, WithThemeProps, withTheme } from '../theming';

interface ButtonProps {
  text?: React.ReactNode;
  iconSource?: ImageSourcePropType;
  component?: React.ReactNode;
}

interface ButtonGroupProps extends TouchableOpacityProps {
  buttons: ButtonProps[];
  selected?: number;
  animationDuration?: number;
  style?: StyleProp<ViewStyle>;
  buttonStyle?: StyleProp<ViewStyle>;
  textStyle?: StyleProp<TextStyle>;
  imageStyle?: StyleProp<ImageStyle>;
  selectorStyle?: StyleProp<ViewStyle>;
  onSelect: Function;
}

const ButtonGroup: React.FC<Themed<typeof createStyleSheet, ButtonGroupProps>> = ({
  styles,
  buttons,
  selected = 0,
  animationDuration = 500,
  activeOpacity = 0.5,
  onSelect,
  style,
  buttonStyle,
  textStyle,
  imageStyle,
  selectorStyle,
}) => {
  const [buttonWidth, setButtonWidth] = useState(0);
  const [animatedValue] = useState(() => new Animated.Value(selected));

  useEffect(() => {
    Animated.timing(animatedValue, {
      duration: animationDuration,
      toValue: selected * buttonWidth,
      useNativeDriver: true,
      easing: Easing.out(Easing.exp),
    }).start();
  }, [animatedValue, selected, buttonWidth, animationDuration]);

  const onLayout = (event: LayoutChangeEvent) => {
    setButtonWidth(event.nativeEvent.layout.width / buttons.length);
  };

  const selectorStyles = [
    styles.selector,
    selectorStyle,
    {
      width: buttonWidth,
      transform: [{ translateX: animatedValue }],
    },
  ];

  const renderButton = (button: ButtonProps, index: number) => {
    if (button.component) {
      return button.component;
    }

    let content = [];

    if (button.iconSource) {
      const imageStyles = [styles.image, imageStyle, index === selected && styles.selectedImage];
      content.push(<Image key="image" style={imageStyles} source={button.iconSource} />);
    }

    if (button.text) {
      const textStyles = [styles.text, textStyle, index === selected && styles.selectedText];
      content.push(
        <Text key="text" style={textStyles} numberOfLines={1}>
          {button.text}
        </Text>,
      );
    }

    return content;
  };

  return (
    <View style={[styles.root, style]}>
      <Animated.View style={selectorStyles} />
      <View style={styles.content} onLayout={onLayout}>
        {buttons.map((button, index) => (
          <TouchableOpacity
            key={index}
            activeOpacity={activeOpacity}
            style={[styles.button, buttonStyle]}
            onPress={() => onSelect(index)}
          >
            {renderButton(button, index)}
          </TouchableOpacity>
        ))}
      </View>
    </View>
  );
};

const createStyleSheet = ({ Colors, Fonts }: Theme) =>
  StyleSheet.create({
    root: {
      height: 40,
      borderWidth: 3,
      borderRadius: 8,
      borderColor: Colors.gray5,
      backgroundColor: Colors.gray5,
    },
    content: {
      flexGrow: 1,
      flexDirection: 'row',
    },
    selector: {
      ...StyleSheet.absoluteFillObject,
      borderRadius: 5,
      backgroundColor: Colors.white,
    },
    button: {
      flex: 1,
      paddingHorizontal: 5,
      alignItems: 'center',
      flexDirection: 'row',
      justifyContent: 'center',
    },
    text: {
      ...Fonts.semibold,
      fontSize: 15,
      marginHorizontal: 5,
      color: Colors.gray,
    },
    image: {
      width: 18,
      height: 18,
      tintColor: Colors.gray,
    },
    selectedText: {
      color: Colors.blue,
    },
    selectedImage: {
      tintColor: Colors.blue,
    },
  });

export type Props = ButtonGroupProps & WithThemeProps;

export default withTheme(ButtonGroup, createStyleSheet, 'ButtonGroup');
