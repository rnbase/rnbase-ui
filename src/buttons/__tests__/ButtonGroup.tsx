import React from 'react';
import { Animated as AnimatedObj, Text } from 'react-native';

// Note: test renderer must be required after react-native.
import TestRenderer, { act } from 'react-test-renderer';

import { Props } from '../ButtonGroup';

jest.doMock('react-native/Libraries/Animated/src/Animated', () => {
  const Animated = jest.requireActual('react-native/Libraries/Animated/src/Animated');

  return {
    ...Animated,
    timing: jest.fn((value: any, { toValue }: any) => {
      value.setValue(toValue);

      return {
        start: jest.fn(),
      };
    }),
  };
});

const TouchableOpacity = (props: any) => React.createElement('TouchableOpacity', props);

jest.doMock('react-native/Libraries/Components/Touchable/TouchableOpacity', () => TouchableOpacity);

/**
 * Under test
 */
const { default: ButtonGroup } = jest.requireActual('../ButtonGroup');

jest.mock('react-native/Libraries/Animated/src/Easing', () => ({
  in: (v: any) => `Easing.in(${v})`,
  out: (v: any) => `Easing.out(${v})`,
  exp: 'Easing.exp',
}));

jest.mock('../Button', () => 'Button');

const Animated: any = AnimatedObj;

const onSelect = jest.fn();
const buttons = [
  {
    text: 'one',
  },
  {
    text: 'two',
  },
];

const createElement = (props: Props) => <ButtonGroup {...props} />;

const createRenderer = (props: Props) => TestRenderer.create(createElement(props));

it('should render normally', () => {
  expect(createRenderer({ buttons, onSelect })).toMatchSnapshot();
});

it('should render icons', () => {
  expect(
    createRenderer({
      buttons: [
        {
          iconSource: { uri: 'icon-one.png' },
        },
        {
          iconSource: { uri: 'icon-two.png' },
        },
      ],
      onSelect,
    }),
  ).toMatchSnapshot();
});

it('should render components', () => {
  expect(
    createRenderer({
      buttons: [
        {
          component: <Text>ONE</Text>,
        },
        {
          component: <Text>TWO</Text>,
        },
      ],
      onSelect,
    }),
  ).toMatchSnapshot();
});

it('should call onSelect', () => {
  const tree = createRenderer({ buttons, onSelect });

  const button = tree.root.findAllByType(TouchableOpacity)[0];

  button.props.onPress();

  expect(onSelect).toBeCalled();
});

describe('selected button animation', () => {
  const start = jest.fn();

  Animated.timing.mockImplementation(() => ({ start }));

  beforeEach(() => {
    start.mockClear();
  });

  it('should not animate', () => {
    const tree = createRenderer({ buttons, onSelect });

    act(() => {});

    Animated.timing.mockClear();

    act(() => {
      tree.update(createElement({ buttons, onSelect, selected: 0 }));
    });

    expect(Animated.timing).not.toBeCalled();
  });

  it('should animate to second button', () => {
    const tree = createRenderer({ buttons, onSelect });

    act(() => {});

    const width = 300;
    const { onLayout } = tree.root.findByProps({ testID: 'container' }).props;

    act(() => {
      onLayout({ nativeEvent: { layout: { width } } });
    });

    Animated.timing.mockClear();

    act(() => {
      tree.update(createElement({ buttons, onSelect, selected: 1 }));
    });

    expect(Animated.timing).toBeCalledWith(expect.objectContaining({ _value: 0 }), {
      toValue: width / buttons.length,
      duration: 500,
      useNativeDriver: true,
      easing: 'Easing.out(Easing.exp)',
    });
    expect(start).toBeCalled();
  });
});
