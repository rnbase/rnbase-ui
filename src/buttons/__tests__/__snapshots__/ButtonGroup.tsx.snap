// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render components 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#e5e5ea",
        "borderColor": "#e5e5ea",
        "borderRadius": 8,
        "borderWidth": 3,
        "height": 40,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#ffffff",
        "borderRadius": 5,
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": Array [
          Object {
            "translateX": 0,
          },
        ],
        "width": 0,
      }
    }
  />
  <View
    onLayout={[Function]}
    style={
      Object {
        "flexDirection": "row",
        "flexGrow": 1,
      }
    }
    testID="container"
  >
    <TouchableOpacity
      activeOpacity={0.5}
      onPress={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "center",
            "paddingHorizontal": 5,
          },
          undefined,
        ]
      }
    >
      <Text>
        ONE
      </Text>
    </TouchableOpacity>
    <TouchableOpacity
      activeOpacity={0.5}
      onPress={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "center",
            "paddingHorizontal": 5,
          },
          undefined,
        ]
      }
    >
      <Text>
        TWO
      </Text>
    </TouchableOpacity>
  </View>
</View>
`;

exports[`should render icons 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#e5e5ea",
        "borderColor": "#e5e5ea",
        "borderRadius": 8,
        "borderWidth": 3,
        "height": 40,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#ffffff",
        "borderRadius": 5,
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": Array [
          Object {
            "translateX": 0,
          },
        ],
        "width": 0,
      }
    }
  />
  <View
    onLayout={[Function]}
    style={
      Object {
        "flexDirection": "row",
        "flexGrow": 1,
      }
    }
    testID="container"
  >
    <TouchableOpacity
      activeOpacity={0.5}
      onPress={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "center",
            "paddingHorizontal": 5,
          },
          undefined,
        ]
      }
    >
      <Image
        source={
          Object {
            "uri": "icon-one.png",
          }
        }
        style={
          Array [
            Object {
              "height": 18,
              "tintColor": "#838387",
              "width": 18,
            },
            undefined,
            Object {
              "tintColor": "#007aff",
            },
          ]
        }
      />
    </TouchableOpacity>
    <TouchableOpacity
      activeOpacity={0.5}
      onPress={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "center",
            "paddingHorizontal": 5,
          },
          undefined,
        ]
      }
    >
      <Image
        source={
          Object {
            "uri": "icon-two.png",
          }
        }
        style={
          Array [
            Object {
              "height": 18,
              "tintColor": "#838387",
              "width": 18,
            },
            undefined,
            false,
          ]
        }
      />
    </TouchableOpacity>
  </View>
</View>
`;

exports[`should render normally 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#e5e5ea",
        "borderColor": "#e5e5ea",
        "borderRadius": 8,
        "borderWidth": 3,
        "height": 40,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#ffffff",
        "borderRadius": 5,
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": Array [
          Object {
            "translateX": 0,
          },
        ],
        "width": 0,
      }
    }
  />
  <View
    onLayout={[Function]}
    style={
      Object {
        "flexDirection": "row",
        "flexGrow": 1,
      }
    }
    testID="container"
  >
    <TouchableOpacity
      activeOpacity={0.5}
      onPress={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "center",
            "paddingHorizontal": 5,
          },
          undefined,
        ]
      }
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "color": "#838387",
              "fontFamily": "System",
              "fontSize": 15,
              "fontWeight": "600",
              "includeFontPadding": false,
              "marginHorizontal": 5,
              "textAlignVertical": "center",
            },
            undefined,
            Object {
              "color": "#007aff",
            },
          ]
        }
      >
        one
      </Text>
    </TouchableOpacity>
    <TouchableOpacity
      activeOpacity={0.5}
      onPress={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "center",
            "paddingHorizontal": 5,
          },
          undefined,
        ]
      }
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "color": "#838387",
              "fontFamily": "System",
              "fontSize": 15,
              "fontWeight": "600",
              "includeFontPadding": false,
              "marginHorizontal": 5,
              "textAlignVertical": "center",
            },
            undefined,
            false,
          ]
        }
      >
        two
      </Text>
    </TouchableOpacity>
  </View>
</View>
`;
